include/pybind11/typing.h

```diff
@@ -0,0 +1,94 @@
+/*
+    pybind11/typing.h: Convenience wrapper classes for basic Python types
+    with more explicit annotations.
+
+    Copyright (c) 2016 Wenzel Jakob <wenzel.jakob@epfl.ch>
+
+    All rights reserved. Use of this source code is governed by a
+    BSD-style license that can be found in the LICENSE file.
+*/
+
+#pragma once
+
+#include "detail/common.h"
+#include "cast.h"
+#include "pytypes.h"
+
+PYBIND11_NAMESPACE_BEGIN(PYBIND11_NAMESPACE)
+
+/*
+    The following types can be used to direct pybind11-generated docstrings
+    to have have more explicit types (e.g., `List[str]` instead of `list`).
+    Just use these in place of existing types.
+
+    There is no additional enforcement of types at runtime.
+*/
+
+template <typename... Types>
```

Comment:
I think this feature will be clearer as such by having the types below in a `typing` sub-namespace.

I'm not sure about (e.g.) `py::typing::Tuple` vs `py::typing::tuple`. Pure Python `typing.tuple` does not exist (I tried with Python 3.11), therefore I'm guessing `py::typing::Tuple` might be less surprising.